File: android/src/main/kotlin/com/kurban/xue_hua_gaode_map/map/OfflineMapHandler.kt
package com.kurban.xue_hua_gaode_map.map

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.amap.api.maps.MapsInitializer
import com.amap.api.maps.offlinemap.OfflineMapCity
import com.amap.api.maps.offlinemap.OfflineMapManager
import com.amap.api.maps.offlinemap.OfflineMapProvince
import com.amap.api.maps.offlinemap.OfflineMapStatus
import com.kurban.xue_hua_gaode_map.AmapPrivacyState
import io.flutter.plugin.common.EventChannel

import java.util.Collections

class OfflineMapHandler(private val context: Context) {
    private var manager: OfflineMapManager? = null
    private var eventSink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private val trackedCityCodes = Collections.synchronizedSet(mutableSetOf<String>())

    fun setEventSink(sink: EventChannel.EventSink?) {
        eventSink = sink
    }

    fun setStoragePath(path: String) {
        MapsInitializer.sdcardDir = path
    }

    private fun ensureManager(): OfflineMapManager {
        AmapPrivacyState.requirePrivacyAgreed()
        if (manager == null) {
            manager = OfflineMapManager(context, object : OfflineMapManager.OfflineMapDownloadListener {
                override fun onDownload(status: Int, completeCode: Int, msg: String?) {
                    val city = resolveProgressCity(msg)
                    city?.code?.let { code ->
                        if (status == OfflineMapStatus.SUCCESS || status == OfflineMapStatus.ERROR) {
                            trackedCityCodes.remove(code)
                        }
                    }
                    val payload = mapOf(
                        "cityCode" to (city?.code ?: ""),
                        "cityName" to (city?.city ?: ""),
                        "status" to statusWire(status),
                        "completePercent" to completeCode,
                        "errorInfo" to msg,
                    )
                    mainHandler.post { eventSink?.success(payload) }
                }

                override fun onCheckUpdate(hasNew: Boolean, name: String?) {}

                override fun onRemove(success: Boolean, name: String?, describe: String?) {}
            })
        }
        return manager!!
    }

    fun getCityList(): List<Map<String, Any?>> {
        val offlineManager = ensureManager()
        val provinces = offlineManager.offlineMapProvinceList ?: return emptyList()
        val result = mutableListOf<Map<String, Any?>>()
        for (province in provinces) {
            result.add(cityToMap(province, isProvince = true))
            province.cityList?.forEach { city ->
                result.add(cityToMap(city, isProvince = false, provinceName = province.provinceName))
            }
        }
        return result
    }

    fun downloadByCityCode(cityCode: String) {
        val city = findCity(cityCode)
            ?: throw IllegalArgumentException("city not found: $cityCode")
        trackedCityCodes.add(city.code)
        ensureManager().downloadByCityCode(city.code)
    }

    fun downloadByCityName(cityName: String) {
        val city = findCityByName(cityName)
            ?: throw IllegalArgumentException("city not found: $cityName")
        trackedCityCodes.add(city.code)
        ensureManager().downloadByCityName(cityName)
    }

    fun pause(cityCode: String) {
        val city = findCity(cityCode)
            ?: throw IllegalArgumentException("city not found: $cityCode")
        ensureManager().pauseByName(city.city)
    }

    fun resume(cityCode: String) {
        downloadByCityCode(cityCode)
    }

    fun remove(cityCode: String) {
        val city = findCity(cityCode)
            ?: throw IllegalArgumentException("city not found: $cityCode")
        ensureManager().remove(city.city)
    }

    fun getDownloadStatus(cityCode: String): Map<String, Any?>? {
        val city = findCity(cityCode) ?: return null
        return cityToMap(city, isProvince = false)
    }

    fun destroy() {
        manager?.destroy()
        manager = null
        eventSink = null
        trackedCityCodes.clear()
    }

    private fun resolveProgressCity(msg: String?): OfflineMapCity? {
        val downloading = manager?.downloadingCityList ?: return null
        if (msg != null) {
            downloading.firstOrNull { it.city == msg || it.code == msg }?.let { return it }
        }
        synchronized(trackedCityCodes) {
            downloading.firstOrNull { it.code in trackedCityCodes }?.let { return it }
        }
        return downloading.firstOrNull()
    }

    private fun findCityByName(cityName: String): OfflineMapCity? {
        val offlineManager = ensureManager()
        offlineManager.offlineMapCityList?.firstOrNull { it.city == cityName }?.let { return it }
        offlineManager.offlineMapProvinceList?.forEach { province ->
            province.cityList?.forEach { city ->
                if (city.city == cityName) return city
            }
        }
        return null
    }

    private fun findCity(cityCode: String): OfflineMapCity? {
        val offlineManager = ensureManager()
        offlineManager.getItemByCityCode(cityCode)?.let { return it }
        offlineManager.offlineMapProvinceList?.forEach { province ->
            province.cityList?.forEach { city ->
                if (city.code == cityCode) return city
            }
        }
        return null
    }

    private fun cityToMap(
        city: OfflineMapCity,
        isProvince: Boolean,
        provinceName: String = "",
    ): Map<String, Any?> {
        return mapOf(
            "name" to city.city,
            "cityCode" to city.code,
            "provinceName" to provinceName,
            "isProvince" to isProvince,
            "downloaded" to (city.state == OfflineMapStatus.SUCCESS),
            "completePercent" to city.getcompleteCode(),
            "status" to statusWire(city.state),
        )
    }

    private fun cityToMap(
        province: OfflineMapProvince,
        isProvince: Boolean,
    ): Map<String, Any?> {
        return mapOf(
            "name" to province.provinceName,
            "cityCode" to province.provinceCode,
            "provinceName" to "",
            "isProvince" to isProvince,
            "downloaded" to (province.state == OfflineMapStatus.SUCCESS),
            "completePercent" to province.completeCode,
            "status" to statusWire(province.state),
        )
    }

    private fun statusWire(status: Int): String {
        return when (status) {
            OfflineMapStatus.LOADING, OfflineMapStatus.UNZIP -> "downloading"
            OfflineMapStatus.SUCCESS -> "finished"
            OfflineMapStatus.PAUSE -> "paused"
            OfflineMapStatus.STOP -> "cancelled"
            OfflineMapStatus.ERROR,
            OfflineMapStatus.EXCEPTION_AMAP,
            OfflineMapStatus.EXCEPTION_NETWORK_LOADING,
            OfflineMapStatus.EXCEPTION_SDCARD,
            OfflineMapStatus.START_DOWNLOAD_FAILED -> "error"
            OfflineMapStatus.WAITING -> "waiting"
            else -> "unknown"
        }
    }
}

File: ios/xue_hua_gaode_map/Sources/xue_hua_gaode_map/OfflineMapHandler.swift
import Flutter
import MAMapKit

final class OfflineMapHandler {
    private var eventSink: FlutterEventSink?

    func setEventSink(_ sink: FlutterEventSink?) {
        eventSink = sink
    }

    func getCityList(result: @escaping FlutterResult) {
        guard requirePrivacy(result: result) else { return }
        guard let offlineMap = MAOfflineMap.shared() else {
            result([])
            return
        }
        var cities: [[String: Any?]] = []
        for province in offlineMap.provinces {
            cities.append(cityMap(province, isProvince: true))
            for case let city as MAOfflineItem in province.cities {
                cities.append(cityMap(city, isProvince: false, provinceName: province.name))
            }
        }
        result(cities)
    }

    func downloadByCityCode(_ cityCode: String, result: @escaping FlutterResult) {
        guard requirePrivacy(result: result) else { return }
        guard let item = findItem(cityCode: cityCode) else {
            result(FlutterError(code: "NOT_FOUND", message: "city not found", details: nil))
            return
        }
        download(item: item)
        result(nil)
    }

    func downloadByCityName(_ cityName: String, result: @escaping FlutterResult) {
        guard requirePrivacy(result: result) else { return }
        guard let item = findItem(cityName: cityName) else {
            result(FlutterError(code: "NOT_FOUND", message: "city not found", details: nil))
            return
        }
        download(item: item)
        result(nil)
    }

    func pause(cityCode: String, result: @escaping FlutterResult) {
        guard requirePrivacy(result: result) else { return }
        guard let item = findItem(cityCode: cityCode) else {
            result(FlutterError(code: "NOT_FOUND", message: "city not found", details: nil))
            return
        }
        MAOfflineMap.shared()?.pause(item)
        result(nil)
    }

    func resume(cityCode: String, result: @escaping FlutterResult) {
        downloadByCityCode(cityCode, result: result)
    }

    func remove(cityCode: String, result: @escaping FlutterResult) {
        guard requirePrivacy(result: result) else { return }
        guard let item = findItem(cityCode: cityCode) else {
            result(FlutterError(code: "NOT_FOUND", message: "city not found", details: nil))
            return
        }
        MAOfflineMap.shared()?.delete(item)
        result(nil)
    }

    func getDownloadStatus(cityCode: String, result: @escaping FlutterResult) {
        guard requirePrivacy(result: result) else { return }
        guard let item = findItem(cityCode: cityCode) else {
            result(nil)
            return
        }
        result(cityMap(item, isProvince: false))
    }

    func destroy() {
        eventSink = nil
    }

    private func download(item: MAOfflineItem) {
        MAOfflineMap.shared()?.downloadItem(
            item,
            shouldContinueWhenAppEntersBackground: true,
            downloadBlock: { [weak self] downloadItem, status, info in
            guard let self, let downloadItem else { return }
            DispatchQueue.main.async {
                self.eventSink?([
                    "cityCode": downloadItem.adcode,
                    "cityName": downloadItem.name,
                    "status": self.statusWire(status),
                    "completePercent": self.completePercent(status: status, info: info),
                    "errorInfo": self.errorInfo(status: status, info: info),
                ])
            }
        })
    }

    private func offlineMap() -> MAOfflineMap? {
        MAOfflineMap.shared()
    }

    private func findItem(cityCode: String) -> MAOfflineItem? {
        guard let offlineMap = offlineMap() else { return nil }
        for province in offlineMap.provinces {
            if province.adcode == cityCode { return province }
            for case let city as MAOfflineItem in province.cities {
                if city.adcode == cityCode { return city }
            }
        }
        for municipality in offlineMap.municipalities {
            if municipality.adcode == cityCode { return municipality }
        }
        for city in offlineMap.cities {
            if city.adcode == cityCode { return city }
        }
        return nil
    }

    private func findItem(cityName: String) -> MAOfflineItem? {
        guard let offlineMap = offlineMap() else { return nil }
        for province in offlineMap.provinces {
            if province.name == cityName { return province }
            for case let city as MAOfflineItem in province.cities {
                if city.name == cityName { return city }
            }
        }
        for municipality in offlineMap.municipalities {
            if municipality.name == cityName { return municipality }
        }
        for city in offlineMap.cities {
            if city.name == cityName { return city }
        }
        return nil
    }

    private func cityMap(_ item: MAOfflineItem, isProvince: Bool, provinceName: String = "") -> [String: Any?] {
        [
            "name": item.name,
            "cityCode": item.adcode,
            "provinceName": provinceName,
            "isProvince": isProvince,
            "downloaded": item.itemStatus == .installed,
            "completePercent": completePercent(item: item),
            "status": statusWire(item.itemStatus),
        ]
    }

    private func completePercent(item: MAOfflineItem) -> Int {
        guard item.size > 0 else { return item.itemStatus == .installed ? 100 : 0 }
        return Int((Double(item.downloadedSize) / Double(item.size)) * 100.0)
    }

    private func completePercent(status: MAOfflineMapDownloadStatus, info: Any?) -> Int {
        guard status == .progress,
              let dict = info as? [String: Any],
              let received = dict[MAOfflineMapDownloadReceivedSizeKey] as? NSNumber,
              let expected = dict[MAOfflineMapDownloadExpectedSizeKey] as? NSNumber,
              expected.intValue > 0 else {
            return status == .finished || status == .completed ? 100 : 0
        }
        return Int((Double(received.intValue) / Double(expected.intValue)) * 100.0)
    }

    private func errorInfo(status: MAOfflineMapDownloadStatus, info: Any?) -> String? {
        guard status == .error else { return nil }
        return (info as? NSError)?.localizedDescription
    }

    private func statusWire(_ status: MAOfflineMapDownloadStatus) -> String {
        switch status {
        case .waiting: return "waiting"
        case .start, .progress, .unzip: return "downloading"
        case .completed, .finished: return "finished"
        case .error: return "error"
        case .cancelled: return "cancelled"
        @unknown default: return "unknown"
        }
    }

    private func statusWire(_ status: MAOfflineItemStatus) -> String {
        switch status {
        case .installed: return "finished"
        case .cached: return "downloading"
        case .expired: return "error"
        case .none: return "unknown"
        @unknown default: return "unknown"
        }
    }

    private func requirePrivacy(result: @escaping FlutterResult) -> Bool {
        guard AmapPrivacyState.privacyAgreed else {
            result(FlutterError(
                code: "PRIVACY_NOT_CONFIGURED",
                message: AmapPrivacyState.privacyError().localizedDescription,
                details: nil
            ))
            return false
        }
        return true
    }
}
